Fix auto exponent issue for torch.pow

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49809 Fix auto exponent issue for torch.pow**

Fixes https://github.com/pytorch/xla/issues/2688 #46936

Differential Revision: [D25724176](https://our.internmc.facebook.com/intern/diff/D25724176)

anjali411 committed:
Fix auto exponent issue for torch.pow

[ghstack-poisoned]

mruberry commented on c10/core/Scalar.cpp:
```diff
@@ -21,4 +21,14 @@ Scalar Scalar::conj() const {
+  } else if (isFloatingPoint()) {
+    return std::log(v.d);
+  } else {
+    return std::log(v.i);
```
Is there a bool case to be worried about here?

anjali411 replied:
no because scalar only stores double, complex, or int. In case a scalar is initialized with bool, it saves a flag indicating it is bool and saves the value as an int

mruberry commented on c10/core/Scalar.h:
```diff
@@ -88,6 +88,18 @@ class C10_API Scalar {
+  Scalar log() const;
+
+  template<typename T>
+  bool equal(T num) const {
```
When comparing a Scalar against a C++ number should people always use equal?

anjali411 replied:
yeah

albanD replied:
I am not very familiar with c++ semantic, why is it different from `operator==`?

anjali411 replied:
To compare a scalar to a number, we first have to obtain the stored value in the scalar which usually is done by `.toDouble()` but with the possibility that the scalar can also be complex, we can't always do that anymore. With the new `equal` method, that's automatically taken care of

mruberry replied:
Offline conversation: add a special case for bool

albanD replied:
Ho you cannot template the `operator==` the same way as you do with equal?

Given that using `.equal()` and not `==` is a (small) constraint I'm just trying to understand why we actually need to do it. But this is not blocking for this PR!

anjali411 replied:
`operator==` can be overloaded to compare two objects of the same type (in this case, scalars), but not objects of different types (a scalar and a number). If we do add this overload, we'd first have to convert the number to a scalar and then compare it with the scalar object in question. But that, to me, seems like an overkill

albanD commented on c10/core/Scalar.h:
```diff
@@ -88,6 +88,18 @@ class C10_API Scalar {
+
+  template<typename T>
+  bool equal(T num) const {
+      if (isComplex()) {
```
nit: alignment looks wrong here?

anjali411 committed:
Update on "Fix auto exponent issue for torch.pow"


Fixes https://github.com/pytorch/xla/issues/2688 #46936

[ghstack-poisoned]

anjali411 committed:
Update on "Fix auto exponent issue for torch.pow"


Fixes https://github.com/pytorch/xla/issues/2688 #46936

[ghstack-poisoned]

mruberry reviewed (approved):
See one inline comment, otherwise looks great!

anjali411 committed:
Update on "Fix auto exponent issue for torch.pow"


Fixes https://github.com/pytorch/xla/issues/2688 #46936

[ghstack-poisoned]

anjali411 committed:
Update on "Fix auto exponent issue for torch.pow"


Fixes https://github.com/pytorch/xla/issues/2688 #46936

[ghstack-poisoned]

mruberry reviewed (changes requested):
@ngimel points out that this can't edit the == for complex because we're trying to be compliant with the standard.

anjali411 committed:
Update on "Fix auto exponent issue for torch.pow"


Fixes https://github.com/pytorch/xla/issues/2688 #46936

Differential Revision: [D25724176](https://our.internmc.facebook.com/intern/diff/D25724176)

[ghstack-poisoned]

anjali411 commented:
> @ngimel points out that this can't edit the == for complex because we're trying to be compliant with the standard.

I removed the == overload for int and c10::complex and only added the comparison functionality for complex scalar and ints and vice versa.

anjali411 committed:
Update on "Fix auto exponent issue for torch.pow"


Fixes https://github.com/pytorch/xla/issues/2688 #46936

Differential Revision: [D25724176](https://our.internmc.facebook.com/intern/diff/D25724176)

[ghstack-poisoned]

ngimel reviewed (approved):
Looks good!

mruberry reviewed (approved):
Looks good; let's wait for the tests to finish.
